www/docs/client/mocking.md

```diff
@@ -0,0 +1,234 @@
+---
+id: Mocking
+title: Mocking your server
+sidebar_label: Mocking your server
+slug: /mocking
+---
+
+Mocking tRPC gives you the ability to isolate your frontend components into testing or presentation environments without having to alter how they behave.
+
+By intercepting our tRPC calls at the [Service Worker level](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) and returning responses as the component expects, mocking can help you to:
+
+- Write performant, easy to execute tests that don't rely on changing the underlying component in any way
+- Present componenets in frontend workshops like Storybook exactly as they would appear and behave in your main app
+
+The examples below use [Jest](https://facebook.github.io/jest/docs/en/tutorial-react.html), (React Testing Library)[https://github.com/testing-library/react-testing-library], and [Mock Service Worker](https://www.npmjs.com/package/msw-trpc) (MSW), but the concepts apply to any testing framework.
+
+## Creating a TRPCMockedProvider
+
+Every test for a React component that uses your `tRPC client` must make it available in React's context.
+
+- In React, you achieve this by wrapping your component tree with the `trpc.Provider` and `QueryClientProvider` components.
+- In Next.js, this step is abstracted in `_app.ts` by using `trpc.withTRPC(MyApp)`.
+
+In tests however, this context is not avaliable by default. Instead, you will need to expose this context by creating a `TRPCMockedProvider`.
+
+### 1. Initialize a new React hook
+
+Create a new file for your `MockedProvider` and import your `AppRouter` type from where your standard tRPC router is defined. This will ensure that your eventual mocking is typed to the standard signature of your entire API.
+
+```ts title='trpcMockedProvider.ts'
+import type { AppRouter } from "../path/to/server/trpc";
+```
+
+Next, create a set of React hooks from your `AppRouter` type signature with `createTRPCReact`, as all the tests and isolated environments we want to render our components in will be in a React environment.
+
+```ts title='trpcMockedProvider.ts'
+const mockedtRPC = createTRPCReact<AppRouter>({
+  unstable_overrides: {
+    useMutation: {
+      async onSuccess(opts) {
+        await opts.originalFn();
+        await opts.queryClient.invalidateQueries();
+      },
+    },
+  },
+});
+```
+
+### 2. Create your mocked tRPC provider
+
+Now we just need to create a wrapper that exposes the client. This process is similar to [the React setup](/reactjs/introduction.mdx) however we don't need to utilize the `useState` hook because our tests are not using SSR.
```

Comment:
Maybe link to the specific equivalent part of the React docs? (https://trpc.io/docs/react#3-add-trpc-providers)